File: frontend/src/api/exportJob.ts
import { get, post } from "./index";
import type { PaginatedData } from "./index";

export type ExportJobStatus = "queued" | "running" | "success" | "failed" | "expired";

export interface ExportJob {
    id: number;
    org_id?: number | null;
    user_id: number;
    operator_name?: string | null;
    module: string;
    export_type: string;
    title: string;
    filename: string;
    status: ExportJobStatus;
    progress: number;
    row_count?: number | null;
    file_size?: number | null;
    error_message?: string | null;
    result_summary?: Record<string, any> | null;
    started_at?: string | null;
    finished_at?: string | null;
    expires_at?: string | null;
    created_at: string;
    updated_at: string;
}

export interface CreateExportJobPayload {
    export_type: string;
    title: string;
    filename: string;
    params: Record<string, any>;
}

export interface ExportJobListParams {
    page?: number;
    page_size?: number;
    status?: ExportJobStatus | "";
    module?: string;
    mine_only?: boolean;
}

export interface ExportDownloadCredential {
    access_key_id: string;
    access_key_secret: string;
    security_token: string;
    expiration: string;
    region: string;
    bucket: string;
    endpoint: string;
    oss_key: string;
    filename: string;
    file_size?: number | null;
}

export function createExportJob(data: CreateExportJobPayload) {
    return post<ExportJob>("/export-jobs", data);
}

export function listExportJobs(params: ExportJobListParams) {
    return get<PaginatedData<ExportJob>>("/export-jobs", params);
}

export function getExportJob(id: number) {
    return get<ExportJob>(`/export-jobs/${id}`);
}

export function getExportJobDownloadCredential(id: number) {
    return post<ExportDownloadCredential>(`/export-jobs/${id}/download-credential`);
}

type OssContent = Blob | ArrayBuffer | Uint8Array | string;
type OssGetResult = { content: OssContent };
type OssClient = {
    get: (name: string, options?: Record<string, any>) => Promise<OssGetResult>;
};

const OSS_DOWNLOAD_TIMEOUT = 2 * 60 * 1000;
const OSS_CHUNKED_DOWNLOAD_THRESHOLD = 32 * 1024 * 1024;
const OSS_DOWNLOAD_CHUNK_SIZE = 8 * 1024 * 1024;
const OSS_DOWNLOAD_PARALLEL = 4;

export async function downloadExportJobFile(id: number) {
    const credential = await getExportJobDownloadCredential(id);
    const { default: OSS } = await import("ali-oss");
    const client = new OSS({
        region: credential.region,
        bucket: credential.bucket,
        endpoint: credential.endpoint,
        accessKeyId: credential.access_key_id,
        accessKeySecret: credential.access_key_secret,
        stsToken: credential.security_token,
        secure: credential.endpoint.startsWith("https://"),
        timeout: OSS_DOWNLOAD_TIMEOUT,
    }) as OssClient;

    const fileSize = credential.file_size || 0;
    const blob =
        fileSize >= OSS_CHUNKED_DOWNLOAD_THRESHOLD
            ? await downloadOssObjectInChunks(client, credential.oss_key, fileSize)
            : await downloadOssObject(client, credential.oss_key);

    saveBlob(blob, credential.filename || "导出文件.xlsx");
}

async function downloadOssObject(client: OssClient, ossKey: string): Promise<Blob> {
    const result = await client.get(ossKey, { timeout: OSS_DOWNLOAD_TIMEOUT });
    return contentToBlob(result.content);
}

async function downloadOssObjectInChunks(client: OssClient, ossKey: string, fileSize: number): Promise<Blob> {
    const chunks = Array.from(
        { length: Math.ceil(fileSize / OSS_DOWNLOAD_CHUNK_SIZE) },
        (_, index) => {
            const start = index * OSS_DOWNLOAD_CHUNK_SIZE;
            const end = Math.min(start + OSS_DOWNLOAD_CHUNK_SIZE, fileSize) - 1;
            return { index, start, end };
        },
    );
    const parts = new Array<Blob>(chunks.length);
    let cursor = 0;

    async function worker() {
        while (cursor < chunks.length) {
            const chunk = chunks[cursor++];
            const result = await client.get(ossKey, {
                timeout: OSS_DOWNLOAD_TIMEOUT,
                headers: {
                    Range: `bytes=${chunk.start}-${chunk.end}`,
                },
            });
            parts[chunk.index] = contentToBlob(result.content);
        }
    }

    const workerCount = Math.min(OSS_DOWNLOAD_PARALLEL, chunks.length);
    await Promise.all(Array.from({ length: workerCount }, () => worker()));
    return new Blob(parts, {
        type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    });
}

function contentToBlob(content: OssContent): Blob {
    if (content instanceof Blob) return content;
    return new Blob([content]);
}

function saveBlob(blob: Blob, filename: string) {
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
}

File: backend/app/models/export_job.py
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, Index, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, SoftDeleteMixin


class ExportJob(SoftDeleteMixin, Base):
    __tablename__ = "fin_export_jobs"
    __table_args__ = (
        Index("idx_fin_export_jobs_user_status", "user_id", "status", "created_at"),
        Index("idx_fin_export_jobs_org_created", "org_id", "created_at"),
        {"comment": "异步导出任务表"},
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True, comment="主键ID")
    org_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("fin_organizations.id"), nullable=True, comment="所属组织ID")
    user_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("fin_users.id"), nullable=False, comment="创建用户ID")
    celery_task_id: Mapped[str | None] = mapped_column(String(200), nullable=True, comment="异步任务ID")

    module: Mapped[str] = mapped_column(String(50), nullable=False, comment="导出模块")
    export_type: Mapped[str] = mapped_column(String(80), nullable=False, comment="导出类型")
    title: Mapped[str] = mapped_column(String(200), nullable=False, comment="任务标题")
    filename: Mapped[str] = mapped_column(String(500), nullable=False, comment="文件名")
    params: Mapped[dict] = mapped_column(JSONB, nullable=False, comment="导出参数")

    status: Mapped[str] = mapped_column(String(20), default="queued", nullable=False, comment="状态")
    progress: Mapped[int] = mapped_column(Integer, default=0, nullable=False, comment="进度百分比")
    row_count: Mapped[int | None] = mapped_column(Integer, nullable=True, comment="导出行数")
    file_size: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="导出文件大小")
    oss_key: Mapped[str | None] = mapped_column(String(1000), nullable=True, comment="OSS文件路径")
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True, comment="错误信息")
    result_summary: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="导出结果摘要")

    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, comment="开始时间")
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, comment="结束时间")
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, comment="过期时间")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), comment="创建时间")
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), comment="更新时间")
